### PR TITLE
Support uploading to multiple Artifactory servers

### LINKED ISSUE #8425: Investigate adding an artifactory server at UNB
With the increasing number of machines located at UNB and the 1G interface for these machines to the outside world there has been failures that suspiciously appear to be bandwidth related when downloading the SDK under test.

This issue is to track discussion on what could be done to reduce the amount of traffic across the network thereby reducing the possibility of network congestion causing download failures.

### PR DESCRIPTION
**Pull request by AdamBrousseau**
Add support for multiple Artifactory servers

Artifactory bandwidth has become an issue in the CI builds.
This has caused issues with upload and download times, as well
as failing downloads due to infrastructure inadequacy.
Particularly for UNB machines, where a large portion of the
farm is located, combined with the pipe into UNB being small.

Being able to setup multiple Artifactory servers spread across
geographies and colocated with pools of machines will allow us 
to push and more importantly pull from a nearby server.

The design leaves one site as the main (default) server (OSU). We
will always upload to the default server. We will only upload
to a secondary server(s) if there are machines with a matching
platform, colocated with another (non-default) server. The design
is laid out so it will scale with more servers without any code
change other than adding to the defaults.yml config.

Since we cannot determine where a test will land (geographically),
we need to upload to any/all servers that are colocated with machines
of matching platform. We also will always pass the default SDK URL.
We will add a curl wrapper on the nodes which have colocated servers.
This wrapper will redirect requests from one server to another. In 
this case redirect OSU requests to UNB. We will also strip off 
the user/password since a) The servers allow anonymous access and
b) The user's api key will be different for every server.

Changes introduced:
- Redesign the Artifactory config in defaults.yml to support multiple
  servers. Identify servers based on geo(graphy) and identify one
  geo as default.
- Redesign the Artifactory variables as a single hashmap. This keeps
  all the config together. Continue to write out the default server 
  values to env, in order for the parent job to continue as-is.
- Move Artifactory setup onto the compile node. This is needed
  in order to determine where we are compiling and where we need
  to upload.
- Add support for uploading to a server behind a vpn. If we aren't 
  compiling behind the vpn, stash the SDK for later when we can
  grab a node in the same geo as the server we need to push to.
  Note: We cannot publish buildInfo to vpn'd servers. Artifacts
  come from nodes but buildInfo comes from Master, so Master needs
  to see the server.
- Add some more details to the Artifacotry README. Also add more 
  info specific to the UNB setup.

Also:
- Unrelated change to stop adding JAVADOC_LIB_URL to the CUSTOMIZED_SDK_URL.
  CUSTOMIZED_SDK_URL is for test and test doesn't need JAVADOC.

[skip ci]
Issue #8425

Signed-off-by: Adam Brousseau <adam.brousseau88@gmail.com>

### 4 use cases
1. Compile & Test with default server (OSU) (eg. zlinux)
1. Compile & Test at UNB (eg. xlinux)
1. Compile at OSU and Test at UNB (eg. plinux)
1. Compile at UNB and Test at OSU (eg. plinux)

### CODE REVIEW

**Comment by AdamBrousseau** [comment]
Jenkins compile zlinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity,extended plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

File: buildenv/jenkins/common/variables-functions.groovy [hunk 1]
```diff
@@ -702,19 +702,151 @@ def set_slack_channel() {
...
+            //println testNodes
+            for (node in testNodes) {
+                def nodeGeo = get_node_geo(node.getLabelString())
+                def nodePlatform = get_node_platform(node.getLabelString())
+                //println nodeGeo
+                //println nodePlatform
+                // Upload if there is a server at geo where there are machines matching our platform.
+                if (nodePlatform == compilePlatform && ARTIFACTORY_CONFIG['geos'].contains(nodeGeo)) {
+                    ARTIFACTORY_CONFIG[nodeGeo]['uploadBool'] = true
+                    //println ARTIFACTORY_CONFIG[nodeGeo]['uploadBool']
```
Comment (vsebe) [inline buildenv/jenkins/common/variables-functions.groovy]: There are multiple nodes per geo. It should set the flag only once (and skip the other nodes with same geo):
```
if (!ARTIFACTORY_CONFIG[nodeGeo].contains('uploadBool') {
    ARTIFACTORY_CONFIG[nodeGeo]['uploadBool'] = true
}
```

File: buildenv/jenkins/common/variables-functions.groovy [hunk 2]
```diff
@@ -702,19 +702,151 @@ def set_slack_channel() {
...
+                    break
+            }
+        }
+    }
+    if (!arch || !baseOS) {
+        echo "WARNING: Unable to determine node arch/os:'${nodeLabels}'"
+        return ''
+    }
+    switch(baseOS) {
+        case 'aix':
```
Comment (vsebe) [inline buildenv/jenkins/common/variables-functions.groovy]: Could you please simplify the switch statement below, maybe use an array for all the matching cases sharing the same code, for example:
```
def nodePlatform
switch(baseOS) {
    case ['aix, 'windows',  'osx', 'zos'] :
        nodePlatform  = baseOS
        break
    case ['linux', 'ubuntu', 'rhel', 'cent', 'sles']:
        ...
        break
    default:
        echo "WARNING: Unknown baseOS:'${baseOS}'"
        nodePlatform  = ''
        break
    }
return nodePlatform 
```

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity zlinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins test sanity zlinux jdk11

**Comment by AdamBrousseau** [comment]
This is ready @vsebe.
Once this is merged I will need to add the curl wrappers to the UNB nodes.
In the meantime I will add the new server's cert to all the nodes' default Java.

File: buildenv/jenkins/common/build.groovy [hunk 3]
```diff
@@ -325,45 +325,48 @@ def archive_sdk() {
...
+                                       "target": "${ARTIFACTORY_CONFIG['uploadDir']}",
                                        "props": "build.buildIdentifier=${BUILD_IDENTIFIER}"]
                     specs.add(javadocSpec)
                 }
                 def uploadFiles = [files : specs]
                 def uploadSpec = JsonOutput.toJson(uploadFiles)
                 upload_artifactory(uploadSpec)
-                env.CUSTOMIZED_SDK_URL = "${ARTIFACTORY_URL}/${ARTIFACTORY_UPLOAD_DIR}${SDK_FILENAME}"
+
+                // Always use the default server link for the descirption and for test.
```
Comment (pshipton) [inline buildenv/jenkins/common/build.groovy]: type here `descirption`

Comment (AdamBrousseau) [reply buildenv/jenkins/common/build.groovy]: fixed

**Comment by AdamBrousseau** [comment]
Jenkins compile xlinux,plinux,zlinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile xlinux,plinux,zlinux,osx jdk11

**Comment by AdamBrousseau** [comment]
Might have an issue with the cert on cent6/x machines.

Jenkins compile xlinux,xlinuxxl jdk8,jdk11

**Comment by AdamBrousseau** [comment]
Seems ok now. Going to try and stress it a bit.

Jenkins compile xlinux,xlinuxxl,plinux,plinuxxl,osx,osxxl jdk8,jdk11

**Comment by pshipton** [comment]
We don't want to stress the UNB plinux box / artifactory too much while we don't have any access to UNB to reboot it if something goes wrong. I discussed with Joe not enabling artifactory at UNB until after the lockdown is lifted, just in case.

**Comment by AdamBrousseau** [comment]
Didn't see any issues on the last build which had 12 compiles. Other than that 1 failure earlier, I have only seen 1 other retry that succeeded on 1st retry attempt.

While I understand we do not want to take down the host, my understanding from Raj, via @jdekonin, was that the issue was related to CPU being overcommited and there was high confidence the issue was resolved. I understand it's mostly you (Pete) feeling the effects of the nightly curl failures so it's up to you to decide if merging this is worth the risk. I'm fine if we block it for now I just don't want it to sit for 6 months and then have a bunch of merge conflicts that may not be trivial.

**Comment by AdamBrousseau** [comment]
Another option would be to merge this without the unb server added to defauilts.yml. That way the code is in but the server isn't enabled.

**Comment by pshipton** [comment]
I'm fine to merge without the unb server added. There aren't any nightly curl failures any more, since the retry problem was resolved. It doesn't seem worth the risk of taking down the host, because if we do take it down we'll be stuck.

**Comment by AdamBrousseau** [comment]
We'll still have 3 OSU plinux but yes.... :D
I will update the change and add a comment.

**Comment by AdamBrousseau** [comment]
jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
@vsebe @pshipton 
Commented out UNB line in defaults. Works as expected
Can this be merged now then?

**Comment by AdamBrousseau** [comment]
Note that the cent6 machines do not use usr/bin/curl they use usr local curl-7

>All machines were using usr/bin/curl except for the 7 cent6-x which were using /usr/local which was a symlink to curl 7.61.1. I left both curl wrapper scripts on the proxy
```
$ curl-usr-bin-curl-link
echo $@ | sed 's#--user .*:.* ##' | xargs /usr/bin/curl --resolve 140-211-168-230-openstack.osuosl.org:443:192.168.10.216
$ curl-usr-local-curl_7_61_1-link
echo $@ | sed 's#--user .*:.* ##' | xargs /usr/local/curl-7.61.1/bin/curl --resolve 140-211-168-230-openstack.osuosl.org:443:192.168.10.216
```

Also note, the wrapper has been updated via #10985

**Pull request merged**